src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs

```diff
@@ -213,68 +199,54 @@ public IEnumerable<ManifestVersionUpdate> CalculateManifestRollbacks(string roll
             if (unrecognizedManifestIds.Any())
             {
                 _reporter.WriteLine(string.Format(LocalizableStrings.RollbackDefinitionContainsExtraneousManifestIds, rollbackDefinitionFilePath, string.Join(" ", unrecognizedManifestIds)).Yellow());
-                manifestRollbacks = manifestRollbacks.Where(rollbackManifest => currentManifestIds.Contains(rollbackManifest.Item1));
+                manifestRollbacks = manifestRollbacks.Where(rollbackManifest => currentManifestIds.Contains(rollbackManifest.Id));
             }
 
-            var manifestUpdates = manifestRollbacks
-                .Select(manifest =>
-                {
-                    var installedManifestInfo = GetInstalledManifestVersion(manifest.id);
-                    return new ManifestVersionUpdate(manifest.id, installedManifestInfo.manifestVersion, installedManifestInfo.sdkFeatureBand.ToString(),
-                        manifest.version, manifest.featureBand.ToString());
-                });
+            var manifestUpdates = manifestRollbacks.Select(manifest =>
+            {
+                var (id, (version, band)) = manifest;
+                var (installedVersion, installedBand) = GetInstalledManifestVersion(id);
+                return new ManifestVersionUpdate(id, installedVersion, installedBand.ToString(), version, band.ToString());
+            });
 
             return manifestUpdates;
         }
 
-
         public async Task<IEnumerable<WorkloadDownload>> GetManifestPackageDownloadsAsync(bool includePreviews, SdkFeatureBand providedSdkFeatureBand, SdkFeatureBand installedSdkFeatureBand)
         {
             var downloads = new List<WorkloadDownload>();
             foreach (var manifest in _workloadResolver.GetInstalledManifests())
             {
                 try
                 {
-                    var packageId = _workloadManifestInstaller.GetManifestPackageId(new ManifestId(manifest.Id), providedSdkFeatureBand);
-
-                    bool success;
-                    // After checking the --sdk-version, check the current sdk band, and then the manifest band in that order
-                    (success, var latestVersion) = await GetPackageVersion(packageId, packageSourceLocation: _packageSourceLocation, includePreview: includePreviews);
-                    if (success)
-                    {
-                        downloads.Add(new WorkloadDownload(manifest.Id, packageId.ToString(), latestVersion.ToString()));
-                    }
-
-                    if (!success && !installedSdkFeatureBand.Equals(providedSdkFeatureBand))
+                    PackageId? providedPackageId = null;
+                    var fallbackFeatureBand = new SdkFeatureBand(manifest.ManifestFeatureBand);
+                    SdkFeatureBand[] bands = [providedSdkFeatureBand, installedSdkFeatureBand, fallbackFeatureBand];
+                    var success = false;
+                    foreach (var band in bands.Distinct())
```

Comment:
Instead of checking if the next band equals the previous one, we just use `Distinct()` so that we're only processing actual unique bands.

Reply:
I could add a comment (or multiple comments) that explains what it does, if you think that'll aid in readability. And this code doesn't result in a perf gain. It just takes 3 duplicate blocks of code and puts them into a loop that runs (potentially) 3 times.

Reply:
As I understand it, Distinct does _not_ guarantee that the order will remain the same. I seem to recall it puts everything into a Dictionary then just randomly pulls things out. In practice, I think it's often in the same order, but I wouldn't rely on that being true here if it's avoidable.

Since there are only three things in the list, we could potentially have a "check if we already processed this one (by checking the previous one(s)) and 'continue' if so"?

Reply:
For the record, I was more confused by the original code than MiYanni's version. If you have three things you want to check, you should do it in a loop, not in nested if statements.

Reply:
> As I understand it, Distinct does not guarantee that the order will remain the same.

This is true... to an extent. I had a long discussion with Smit as our resident expert on this kind of stuff. `Distinct` is *declared* to not preserve order but in operation on something like an array or list, it does based on how it is implemented. Smit wrote on the whiteboard right behind my desk a LINQ expression that would solve this were you make sure to order the elements by index in a `Select` that would do the `Distinct` call. But that's a lot of extra work and not worth the effort. The likelihood of the underlying operation for `Distinct` on an array/list changing is very, very low, as it would be a massive breaking change in the runtime.

> "check if we already processed this one (by checking the previous one(s)) and 'continue' if so"?

You've missed the multiple evolutions of this method, but I did do that previously. The problem with that is it took me *quite a while* to realize that was the purpose of the `!fallbackFeatureBand.Equals(installedSdkFeatureBand)` checks in the original code. It wasn't blatantly obvious that was the point of those checks is to not duplicate work on already processed bands. I feel `Distinct` does that job very concisely. However, I was planning on also adding some comments here based on Noah's feedback.

After some discussion with Smit, he gave me a suggestion on another way to do this that should make the clarity of `success` and preserve order. It doesn't use a loop, so I'm going to give that a try now.

Reply:
Oh man. This version of it is so clean... but is has one large caveat. It has no way to deduplicate the bands if they're identical bands! 💢 It would add a bunch of goofy boilerplate to make it do that. This would definitely have a perf penalty but I like it so much! 😭

```C#
        public async Task<IEnumerable<WorkloadDownload>> GetManifestPackageDownloadsAsync(bool includePreviews, SdkFeatureBand providedSdkFeatureBand, SdkFeatureBand installedSdkFeatureBand)
        {
            var downloads = new List<WorkloadDownload>();
            PackageId? providedPackageId = null;
            foreach (var manifest in _workloadResolver.GetInstalledManifests())
            {
                try
                {
                    var fallbackSdkFeatureBand = new SdkFeatureBand(manifest.ManifestFeatureBand);
                    // Try each feature band for acquiring the workload download info.
                    WorkloadDownload workloadDownload =
                        await CreateWorkloadDownload(manifest.Id, providedSdkFeatureBand) ??
                        await CreateWorkloadDownload(manifest.Id, installedSdkFeatureBand) ??
                        await CreateWorkloadDownload(manifest.Id, fallbackSdkFeatureBand);

                    // If we couldn't create the workload download, report the error.
                    if (workloadDownload == null)
                    {
                        _reporter.WriteLine(LocalizableStrings.ManifestPackageUrlNotResolved, providedPackageId);
                    }
                    else
                    {
                        downloads.Add(workloadDownload);
                    }
                }
                catch
                {
                    _reporter.WriteLine(LocalizableStrings.ManifestPackageUrlNotResolved, manifest.Id);
                }
            }

            return downloads;

            async Task<WorkloadDownload> CreateWorkloadDownload(string manifestId, SdkFeatureBand band)
            {
                var packageId = _workloadManifestInstaller.GetManifestPackageId(new ManifestId(manifestId), band);
                providedPackageId ??= packageId;

                try
                {
                    var latestVersion = await _nugetPackageDownloader.GetLatestPackageVersion(packageId, _packageSourceLocation, includePreviews);
                    return new WorkloadDownload(manifestId, packageId.ToString(), latestVersion.ToString());
                }
                catch (NuGetPackageNotFoundException)
                {
                }

                return null;
            }
        }
```

Reply:
After further investigation and trial/error, I'm gonna keep this current implementation with the `foreach` and `Distinct` for a couple of reasons:
- `Distinct` is [currently implemented](https://github.com/dotnet/runtime/blob/f20b2b50f159e64ec45d65e0dc0e7c644ee75b37/src/libraries/System.Linq/src/System/Linq/Distinct.cs#L122-L130) to preserve order. It uses a backing `HashSet` to determine if a value is unique and only returns an element if it was able to be added to the `HashSet`. It uses the iterator of the collection passed to it, so if that collection iterates in order (here, the `Array` iterates in order), then it will always return the elements in order. [Here](https://stackoverflow.com/a/69527769/294804) is a simple explanation of that.
- This actually avoids a perf issue with the previous implementation. The previous implementation worked this way:

| Variable | Value
| --- | ---
| A | 1
| B | 2
| C | 1

Using the little chart here, we know there were 3 bands being checked; I'm using A, B, C, to represent those. It would run `A` and if `A` failed, it would check if `B == A`. If it didn't, it would run `B`. If `B` failed, it would check if `C == B`. If it didn't, then run `C`. The issue here is that it never checks if `C == A`. So it ends up running the **same band twice**. If we use `Distinct`, this scenario can't happen.

Reply:
I've added some code comments to this method that should hopefully aid in understanding. I've also updated `UpdateAdvertisingManifestAsync` to follow this same pattern since it was using the previous pattern except with 2 bands instead of 3.

Reply:
Thank you, looking at it again I much prefer the new version, plus it avoids the A == C problem is quite good, I do believe its a problem because the fallback feature band is the manifest feature band, which could be the same as the current sdk feature band

Reply:
Also the temptation for the clean version is real but I think what you settled with is for the best 😄